File: bugsnag-android-core/src/main/java/com/bugsnag/android/DefaultDelivery.java
package com.bugsnag.android;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.NonNull;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.Map;

public class DefaultDelivery implements Delivery {

    private static final int HTTP_REQUEST_FAILED = 0;
    private final Connectivity connectivity;
    private final Context androidContext;

    DefaultDelivery(Connectivity connectivity, Context androidContext) {
        this.connectivity = connectivity;
        this.androidContext = androidContext;
    }

    @Override
    public void deliver(@NonNull SessionTrackingPayload payload,
                        @NonNull Configuration config) throws DeliveryFailureException {
        String endpoint = config.getSessionEndpoint();
        int status = deliver(endpoint, payload, config.getSessionApiHeaders());

        if (status != 202) {
            Logger.warn("Session API request failed with status " + status, null);
        } else {
            Logger.info("Completed session tracking request");
        }
    }

    @Override
    public void deliver(@NonNull Report report,
                        @NonNull Configuration config) throws DeliveryFailureException {
        String endpoint = config.getEndpoint();
        int status = deliver(endpoint, report, config.getErrorApiHeaders());

        if (status / 100 != 2) {
            Logger.warn("Error API request failed with status " + status, null);
        } else {
            Logger.info("Completed error API request");
        }
    }

    int deliver(String urlString,
                JsonStream.Streamable streamable,
                Map<String, String> headers) throws DeliveryFailureException {

        Logger.setEnabled(true);
        if (urlString.contains("notify.bugsnag.com")) {
            String endpointOverride;
            try {
                // in <application> you set meta data to override the native crash endpoint
                ApplicationInfo info = androidContext.getPackageManager().getApplicationInfo(androidContext.getPackageName(), PackageManager.GET_META_DATA);
                Bundle bundle = info.metaData;
                endpointOverride = bundle.getString("g4g.bugsnag_endpoint");
                Log.i("Bugsnag-g4g", "overriding endpoint to" + endpointOverride);
                urlString = endpointOverride;
            } catch (PackageManager.NameNotFoundException e) {
                Log.wtf("Bugsnag-g4g", e);
            }
            Log.i("Bugsnag-g4g", "overrided endpoint? -> " + urlString);
        }
        if (connectivity != null && !connectivity.hasNetworkConnection()) {
            throw new DeliveryFailureException("No network connection available", null);
        }
        HttpURLConnection conn = null;

        try {
            URL url = new URL(urlString);
            conn = (HttpURLConnection) url.openConnection();
            conn.setDoOutput(true);
            conn.setChunkedStreamingMode(0);
            conn.addRequestProperty("Content-Type", "application/json");

            for (Map.Entry<String, String> entry : headers.entrySet()) {
                conn.addRequestProperty(entry.getKey(), entry.getValue());
            }

            JsonStream stream = null;

            try {
                OutputStream out = conn.getOutputStream();
                Charset charset = Charset.forName("UTF-8");
                BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(out, charset));
                stream = new JsonStream(writer);
                streamable.toStream(stream);
            } finally {
                IOUtils.closeQuietly(stream);
            }

            // End the request, get the response code
            return conn.getResponseCode();
        } catch (IOException exception) {
            throw new DeliveryFailureException("IOException encountered in request", exception);
        } catch (Exception exception) {
            Logger.warn("Unexpected error delivering payload", exception);
            return HTTP_REQUEST_FAILED;
        } finally {
            IOUtils.close(conn);
        }
    }

}

File: bugsnag-android-core/src/main/java/com/bugsnag/android/DefaultDelivery.kt
package com.bugsnag.android

import android.net.TrafficStats
import com.bugsnag.android.internal.HEADER_BUGSNAG_INTEGRITY
import com.bugsnag.android.internal.JsonHelper
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.zip.GZIPOutputStream

internal class DefaultDelivery(
    private val connectivity: Connectivity?,
    private val logger: Logger
) : Delivery {

    override fun deliver(payload: Session, deliveryParams: DeliveryParams): DeliveryStatus {
        val status = deliver(
            deliveryParams.endpoint,
            JsonHelper.serialize(payload),
            payload.integrityToken,
            deliveryParams.headers,
            deliveryParams.payloadEncoding
        )

        logger.i("Session API request finished with status $status")
        return status
    }

    override fun deliver(payload: EventPayload, deliveryParams: DeliveryParams): DeliveryStatus {
        val json = payload.trimToSize().toByteArray()
        val status = deliver(
            deliveryParams.endpoint,
            json,
            payload.integrityToken,
            deliveryParams.headers,
            deliveryParams.payloadEncoding
        )

        logger.i("Error API request finished with status $status")
        return status
    }

    fun deliver(
        urlString: String,
        json: ByteArray,
        integrity: String?,
        headers: Map<String, String?>,
        encoding: DeliveryParams.PayloadEncoding = DeliveryParams.PayloadEncoding.NONE
    ): DeliveryStatus {

        TrafficStats.setThreadStatsTag(1)
        if (connectivity != null && !connectivity.hasNetworkConnection()) {
            return DeliveryStatus.UNDELIVERED
        }
        var conn: HttpURLConnection? = null

        val status = try {
            conn = makeRequest(URL(urlString), json, integrity, headers, encoding)

            // End the request, get the response code
            val responseCode = conn.responseCode
            val status = DeliveryStatus.forHttpResponseCode(responseCode)
            logRequestInfo(responseCode, conn, status)
            status
        } catch (oom: OutOfMemoryError) {
            // attempt to persist the payload on disk. This approach uses streams to write to a
            // file, which takes less memory than serializing the payload into a ByteArray, and
            // therefore has a reasonable chance of retaining the payload for future delivery.
            logger.w("Encountered OOM delivering payload, falling back to persist on disk", oom)
            DeliveryStatus.UNDELIVERED
        } catch (exception: IOException) {
            logger.w("IOException encountered in request", exception)
            DeliveryStatus.UNDELIVERED
        } catch (exception: Exception) {
            logger.w("Unexpected error delivering payload", exception)
            DeliveryStatus.FAILURE
        } finally {
            conn?.disconnect()
        }

        return status
    }

    private fun makeRequest(
        url: URL,
        json: ByteArray,
        integrity: String?,
        headers: Map<String, String?>,
        encoding: DeliveryParams.PayloadEncoding
    ): HttpURLConnection {
        val conn = url.openConnection() as HttpURLConnection
        conn.doOutput = true

        integrity?.let { digest ->
            conn.addRequestProperty(HEADER_BUGSNAG_INTEGRITY, digest)
        }
        headers.forEach { (key, value) ->
            if (value != null) {
                conn.addRequestProperty(key, value)
            }
        }

        if (encoding == DeliveryParams.PayloadEncoding.GZIP) {
            // we don't use fixedLengthStreamingMode to avoid making yet-another gzipped copy

            conn.addRequestProperty("Content-Encoding", "gzip")
            GZIPOutputStream(conn.outputStream).use {
                it.write(json)
            }
        } else {
            // avoids creating a buffer within HttpUrlConnection, see
            // https://developer.android.com/reference/java/net/HttpURLConnection
            conn.setFixedLengthStreamingMode(json.size)

            // write the JSON payload
            conn.outputStream.use {
                it.write(json)
            }
        }
        return conn
    }

    private fun logRequestInfo(code: Int, conn: HttpURLConnection, status: DeliveryStatus) {
        runCatching {
            logger.i(
                "Request completed with code $code, " +
                    "message: ${conn.responseMessage}, " +
                    "headers: ${conn.headerFields}"
            )
        }
        runCatching {
            conn.inputStream.bufferedReader().use {
                logger.d("Received request response: ${it.readText()}")
            }
        }

        runCatching {
            if (status != DeliveryStatus.DELIVERED) {
                conn.errorStream.bufferedReader().use {
                    logger.w("Request error details: ${it.readText()}")
                }
            }
        }
    }
}
